File: 백준/Gold/2313. 보석 구매하기/보석 구매하기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int ans = 0;
        int[][] res = new int[n][2];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            int l = Integer.parseInt(st.nextToken());
            int[] arr = new int[l];
            int[] dp = new int[l + 1];
            int total = Integer.MIN_VALUE;
            int start = 0;
            int end = 0;
            int count = Integer.MAX_VALUE;
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < l; j++) {
                arr[j] = Integer.parseInt(st.nextToken());
                dp[j + 1] = dp[j] + arr[j];
            }
            for (int j = 0; j < l + 1; j++) {
                for (int k = j + 1; k < l + 1; k++) {
                    int tmp = dp[k] - dp[j];
                    if (tmp > total) {
                        total = tmp;
                        start = j + 1;
                        end = k;
                        count = end - start;
                    } else if (tmp == total) {
                        int count1 = k - j - 1;
                        if (count1 < count) {
                            count = count1;
                            start = j + 1;
                            end = k;
                        }
                    }
                }
            }
            ans += total;
            res[i][0] = start;
            res[i][1] = end;
        }
        System.out.println(ans);
        for (int i = 0; i < n; i++) {
            System.out.print(res[i][0]);
            System.out.print(" ");
            System.out.println(res[i][1]);
        }
    }
}

File: 백준/Gold/2313. 보석 구매하기/보석 구매하기.py
import sys
input = sys.stdin.readline

n = int(input())
ans = 0
res = []
for _ in range(n):
    l = int(input())
    arr = list(map(int, input().split()))
    dp = [0] * (l + 1)
    total, start, end, count = -1e11, 0, 0, 0
    for i in range(l):
        dp[i + 1] = dp[i] + arr[i]
    for i in range(l + 1):
        for j in range(i + 1, l + 1):
            tmp = dp[j] - dp[i]
            if tmp > total:
                total = tmp
                start = i + 1
                end = j
                count = end - start
            elif tmp == total:
                count1 = j - i - 1
                if count1 < count:
                    total = tmp
                    start = i + 1
                    end = j
                    count = end - start
    ans += total
    res.append([start, end])
print(ans)
for i in range(n):
    print(*res[i])